web/ui/react-app/src/pages/graph/HistogramChart.tsx

```diff
@@ -31,62 +98,163 @@ const HistogramChart: FC<{ histogram: Histogram; index: number; scale: ScaleType
+
+          {buckets && (
+            <RenderHistogramBars
+              buckets={buckets}
+              scale={scale}
+              rangeMin={rangeMin}
+              rangeMax={rangeMax}
+              index={index}
+              fds={fds}
+              fdMax={fdMax}
+              countMax={countMax}
+              bw={expBucketWidth}
+              maxPositive={maxPositive}
+              startPositive={startPositive}
+              startNegative={startNegative}
+              endPositive={endPositive}
+              widthNegative={widthNegative}
+              widthTotal={widthTotal}
+            />
+          )}
+
           <div className="histogram-axes"></div>
         </div>
         <div className="histogram-x-labels">
-          <div key={0} className="histogram-x-label" style={{ width: 0 }}>
-            0
+          <div className="histogram-x-label">
+            <React.Fragment>
+              <div style={{ position: 'absolute', left: 0 }}>{formatter.format(rangeMin)}</div>
+              {rangeMin < 0 && <div style={{ position: 'absolute', left: zeroAxisLeft }}>0</div>}
+              <div style={{ position: 'absolute', right: 0 }}>{formatter.format(rangeMax)}</div>
+            </React.Fragment>
           </div>
-          {xLabelTicks.map((i) => (
-            <div key={i} className="histogram-x-label">
-              <div style={{ position: 'absolute', right: i === 1 ? 0 : -18 }}>{formatter.format(rangeMax * i)}</div>
-            </div>
-          ))}
         </div>
       </div>
     </div>
   );
 };
 
+interface RenderHistogramProps {
+  buckets: [number, string, string, string][];
+  scale: ScaleType;
+  rangeMin: number;
+  rangeMax: number;
+  index: number;
+  fds: number[];
+  fdMax: number;
+  countMax: number;
+  bw: number;
+  maxPositive: number;
+  startPositive: number;
+  startNegative: number;
+  endPositive: number;
+  widthNegative: number;
+  widthTotal: number;
+}
+
+const RenderHistogramBars: FC<RenderHistogramProps> = ({
+  buckets,
+  scale,
+  rangeMin,
+  rangeMax,
+  index,
+  fds,
+  fdMax,
+  countMax,
+  bw,
+  maxPositive,
+  startPositive,
+  startNegative,
+  endPositive,
+  widthNegative,
+  widthTotal,
+}) => {
+  return (
+    <React.Fragment>
+      {buckets.map((b, bIdx) => {
+        const left = parseFloat(b[1]);
+        const right = parseFloat(b[2]);
+        const count = parseFloat(b[3]);
+        console.log('new stuff');
+
+        const bucketIdx = `bucket-${index}-${bIdx}-${Math.ceil(parseFloat(b[3]) * 100)}`;
+        const bucketWidth =
+          scale === 'linear' ? ((right - left) / (rangeMax - rangeMin)) * 100 + '%' : (bw / widthTotal) * 100 + '%';
+        const bucketLeft =
+          scale === 'linear'
+            ? ((left - rangeMin) / (rangeMax - rangeMin)) * 100 + '%'
+            : left < 0
+            ? (Math.log(Math.abs(left)) - startNegative / widthTotal) * 100 + '%' // negative buckets boundary
```

Comment:
With recent updates and fixed parenthesis:

```suggestion
            ? -(Math.log(Math.abs(left)) - startNegative) / widthTotal * 100 + '%' // negative buckets boundary
```